2274 Benchmark User Guide

Shivansh20128 opened the pull request:
<!--
⚠️ Your pull request title should be short, comprehensive, and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.
-->

<!--
If the validation checks fail
  1. Run `make check-types` (from the root directory of the repository) and fix any mypy (https://mypy.readthedocs.io/en/stable/) errors.
  2. Run `make format` to fix any linting/formatting errors. There may be some issues that require manual intervention for you to fix.

For more information, check the Mitiq style guidelines (https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).
-->

## Description
This PR adds a user guide section for benchmarks.
Closes issue #2274 
<!-- Please explain the changes you made here. -->

---

### License

- [ ] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.

- [ ] I added unit tests for new code.
- [ ] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [ ] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [ ] I [updated the documentation](../blob/main/docs/CONTRIBUTING_DOCS.md) where relevant.
- [ ] Added myself / the copyright holder to the AUTHORS file

Shivansh20128 commented:
Hi! Can someone review this draft PR and reply if this is the kind of documentation the project needs? I will then add the rest of it accordingly. 
Thank You

natestemen commented:
@Shivansh20128 this is exactly the kind of thing we're looking for!

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,62 @@
+        .real
+    )
+
+circuits = benchmarks.generate_ghz_circuit(n_qubits=10)
```
This function returns just 1 GHZ circuit.

Shivansh20128 commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,62 @@
+
+## GHZ Circuits
+
+The GHZ (Greenberger–Horne–Zeilinger) circuits create the GHZ states at are highly sensitive to noise. Thus, they make it easy to test error rates in entanglement creation and preservation, which is central for many quantum algorithms.
```
Okay thanks for the insight!

Shivansh20128 commented on docs/source/guide/benchmarks.md:
````diff
@@ -0,0 +1,62 @@
+
+The GHZ (Greenberger–Horne–Zeilinger) circuits create the GHZ states at are highly sensitive to noise. Thus, they make it easy to test error rates in entanglement creation and preservation, which is central for many quantum algorithms.
+
+```{code-cell} ipython3
````
Yes, I was thinking that too. I will add the main code block before the circuits to be followed as a template, and then only add the circuit function call in each of the subsections.

Shivansh20128 commented:
@purva-thakre Can you tell me why this error is coming in the checks performed? This is a different one from the previous check error, which I have linked to an open issue. I don't think its related to the changes I have made, so maybe I am just unlucky to encounter two tests failing non-deterministically🥲.

cosenal commented:
@Shivansh20128 https://github.com/unitaryfund/mitiq/actions/runs/11801809295/job/32876002160?pr=2566 is definitely unrelated to your changes, but it's not even a test that fails non-deterministically. It looks like Python panicked when calling some core routine written in Rust 😱 Interesting, but definitely not in the scope of this PR 😄

Shivansh20128 commented:
> It looks like Python panicked when calling some core routine written in Rust 😱 Interesting, but definitely not in the scope of this PR 😄

Aahh. Okay. Thank you

Shivansh20128 commented:
I think the error in the docs-build will be resolved once #2570 is merged.

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## Mirror Circuits
+
+The {func}`mitiq.mirror_circuits.generate_mirror_circuit` involves running a quantum circuit forward and then “mirroring” it (applying the reverse operations). Ideally, this results in returning the system to the initial state, so they’re great for testing if the noise mitigation is effective in preserving information through complex sequences.
```
```suggestion
The {func}`.generate_mirror_circuit` involves running a quantum circuit forward and then “mirroring” it (applying the reverse operations). Ideally, this results in returning the system to the initial state, so they’re great for testing if the noise mitigation is effective in preserving information through complex sequences.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## Mirror Quantum Volume Circuits
+
+The {func}`mitiq.mirror_qv_circuits.generate_mirror_qv_circuit` is designed to test `quantum volume`, a metric combining circuit depth, number of qubits, and fidelity. These circuits check whether error mitigation techniques help achieve higher effective quantum volumes on noisy devices.
```
```suggestion
The {func}`.generate_mirror_qv_circuit` is designed to test `quantum volume`, a metric combining circuit depth, number of qubits, and fidelity. These circuits check whether error mitigation techniques help achieve higher effective quantum volumes on noisy devices.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## Quantum Phase Estimation Circuits
+
+The {func}`mitiq.qpe_circuits.generate_qpe_circuit` is used to the measure eigenvalues of unitary operators. Since accurate phase estimation requires precise control over operations, these circuits test the mitigation techniques’ ability to handle small noise effects over multiple gate sequences.
```
```suggestion
The {func}`.generate_qpe_circuit` is used to the measure eigenvalues of unitary operators. Since accurate phase estimation requires precise control over operations, these circuits test the mitigation techniques’ ability to handle small noise effects over multiple gate sequences.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## Quantum Volume Circuits
+
+The {func}`mitiq.quantum_volume_circuits.generate_quantum_volume_circuit` tests the maximum achievable "volume" or computational capacity of a quantum processor. Running these circuits with error mitigation tests if mitiq’s techniques improve the effective quantum volume.
```
```suggestion
The {func}`.generate_quantum_volume_circuit` tests the maximum achievable "volume" or computational capacity of a quantum processor. Running these circuits with error mitigation tests if mitiq’s techniques improve the effective quantum volume.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## Randomized Benchmarking Circuits
+
+The {func}`mitiq.randomized_benchmarking.generate_rb_circuits` are sequences of random gates (generally Clifford gates), to estimate an average error rate. They’re standard in benchmarking for evaluating how well mitiq’s error mitigation reduces this error rate across different levels of noise.
```
```suggestion
The {func}`.generate_rb_circuits` are sequences of random gates (generally Clifford gates), to estimate an average error rate. They’re standard in benchmarking for evaluating how well mitiq’s error mitigation reduces this error rate across different levels of noise.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## Rotated Randomized Benchmarking Circuits
+
+The {func}`mitiq.rotated_randomized_benchmarking.generate_rotated_rb_circuits` are sequences of random gates similar to {func}`mitiq.randomized_benchmarking.generate_rb_circuits`, but with rotations added, that allows assessment of errors beyond just the standard Clifford gates. They’re useful to check how well Mitiq handles noise in scenarios with more diverse gates.
```
```suggestion
The {func}`.generate_rotated_rb_circuits` are sequences of random gates similar to {func}`mitiq.randomized_benchmarking.generate_rb_circuits`, but with rotations added, that allows assessment of errors beyond just the standard Clifford gates. They’re useful to check how well Mitiq handles noise in scenarios with more diverse gates.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## Randomized Clifford+T Circuits
+
+The {func} `mitiq.randomized_clifford_t_circuit.generate_random_clifford_t_circuit` add the T gate to the standard Clifford set, adding more complex operations to the random benchmarking. This type evaluates Mitiq’s performance with gate sets that go beyond the Clifford gates, crucial for fault-tolerant computing.
```
```suggestion
The {func}`.generate_random_clifford_t_circuit` add the T gate to the standard Clifford set, adding more complex operations to the random benchmarking. This type evaluates Mitiq’s performance with gate sets that go beyond the Clifford gates, crucial for fault-tolerant computing.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,116 @@
+
+## W State Circuits
+
+The {func}`mitiq.w_state_circuits.generate_w_circuit` are entangled circuits that distribute the entanglement across qubits differently than GHZ states. Testing with W state circuits can help explore how well a device maintains distributed entanglement in noisy environments.
```
```suggestion
The {func}`.generate_w_circuit` are entangled circuits that distribute the entanglement across qubits differently than GHZ states. Testing with W state circuits can help explore how well a device maintains distributed entanglement in noisy environments.
```

purva-thakre commented on docs/source/guide/benchmarks.md:
````diff
@@ -0,0 +1,116 @@
+
+The {func}`mitiq.generate_ghz_circuit` create the GHZ states that are highly sensitive to noise. Thus, they make it easy to test error rates in entanglement creation and preservation, which is central for many quantum algorithms.
+
+```{code-cell} ipython3
+circuit = benchmarks.generate_ghz_circuit(n_qubits=10)
+```
````
```suggestion
```{code-cell} ipython3
circuit = benchmarks.generate_ghz_circuit(n_qubits=10)

print(circuit)
```
```
It might be good to show what each benchmarking circuit is supposed to look like. If the circuit output is too big, reduce the number of qubits etc.

Shivansh20128 replied:
I have added the suggestion for all the examples.

purva-thakre commented on docs/source/guide/benchmarks.md:
````diff
@@ -0,0 +1,116 @@
+The {func}`mitiq.qpe_circuits.generate_qpe_circuit` is used to the measure eigenvalues of unitary operators. Since accurate phase estimation requires precise control over operations, these circuits test the mitigation techniques’ ability to handle small noise effects over multiple gate sequences.
+
+```{code-cell} ipython3
+circuit = benchmarks.qpe_circuits.generate_qpe_circuit(evalue_reg=3)
````
```suggestion
from mitiq.benchmarks import generate_qpe_circuit

circuit = generate_qpe_circuit(evalue_reg=3)
```

explicitly import each function `from mitiq.benchmarks import generate_qpe_circuit` when you create that circuit.

Shivansh20128 replied:
I have added this suggestion to all the examples.

Shivansh20128 commented:
What are these errors regarding? It mentions something about it being unable to find the docker images locally.

purva-thakre commented:
> What are these errors regarding? It mentions something about it being unable to find the docker images locally.

Which ones?

Shivansh20128 commented:
> Which ones?

Hmm they have stopped now. You can see it [here](https://github.com/unitaryfund/mitiq/actions/runs/11820643050/job/32933428735) if you want.

purva-thakre commented on docs/source/guide/benchmarks.md:
````diff
@@ -0,0 +1,138 @@
+```{code-cell} ipython3
+from mitiq.benchmarks import generate_ghz_circuit
+
+circuit = generate_ghz_circuit(n_qubits=10)
````
reduce the number of qubits. Right now, the output circuit is pretty big. 

Apply the same idea to other example circuits on this page.

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,138 @@
+
+## Randomized Benchmarking Circuits
+
+The {func}`.generate_rb_circuits` are sequences of random gates (generally Clifford gates), to estimate an average error rate. They’re standard in benchmarking for evaluating how well mitiq’s error mitigation reduces this error rate across different levels of noise.
```
Can you add some details about what the state of each circuit is supposed to be? For example, the ideal output of a randomized benchmarking circuit is an identity. 

GHZ and W circuits define a state. So, you could use some latex equations to show what the states are supposed to be. 

We don't want to describe what the addition of noise will do to a circuit. the focus is on the details of the benchmarking circuit. All docstrings link a reference used to define a particular benchmarking circuit. You could use details from these refs if you don't know a lot about these circuits.

purva-thakre replied:
You can also use [`.final_state_vector()`](https://quantumai.google/reference/python/cirq/final_state_vector) to show the ideal circuit state, which is what we expect it to be. 

```python
from mitiq.benchmarks import generate_ghz_circuit
circuit = generate_ghz_circuit(n_qubits=3)
circuit.final_state_vector()
```
this will spit out the 1-D vector form of a 3 qubit GHZ state. 

![image](https://github.com/user-attachments/assets/0d0f16a4-481b-4fe1-b90f-2d23ddfdd6e8)

Shivansh20128 replied:
> You can also use [`.final_state_vector()`](https://quantumai.google/reference/python/cirq/final_state_vector) to show the ideal circuit state, which is what we expect it to be.

Yes, I can add this function to the example workflow at the top of the file, so every circuit's ideal circuit state can be seen.

Shivansh20128 replied:
> We don't want to describe what the addition of noise will do to a circuit. the focus is on the details of the benchmarking circuit. All docstrings link a reference used to define a particular benchmarking circuit. You could use details from these refs if you don't know a lot about these circuits.

Since the section is about benchmarking, I believe we should show improvements with mitiq when compared to a noisy circuit without mitiq. But if you think otherwise, I can remove that part from the examples.

Also, since I have now added [.final_state_vector()](https://quantumai.google/reference/python/cirq/final_state_vector), shall I still go ahead and add the latex equations to show what the states are supposed to be?

purva-thakre replied:
The focus is on benchmarking circuits. Not benchmarking the performance of a QEM technique using mitiq.

cosenal commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,194 @@
+
+# Benchmarks
+
+Mitiq benchmarks error mitigation techniques by running quantum circuits with and without mitigation, measuring improvements in accuracy, fidelity, and error rates. The process involves executing various circuit types—like GHZ, Mirror, Quantum Volume, and Randomized Benchmarking circuits—and comparing mitigated results against ideal outcomes. Analysis of these benchmarking results produces perfor ...
```
Just saying `accuracy, fidelity, and error rates` sounds a bit too informal. We could be more specific about defining the metrics that one can evaluate from a QEM technique, that is, defining _fidelity_ more formally.
Something else that one wants to evaluate is the overhead cost of a technique, e.g., how big the intermediate circuits are in ZNE compared to PEC. We should mention overhead as one of the benchmark metrics.

cosenal commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,194 @@
+
+Mitiq benchmarks error mitigation techniques by running quantum circuits with and without mitigation, measuring improvements in accuracy, fidelity, and error rates. The process involves executing various circuit types—like GHZ, Mirror, Quantum Volume, and Randomized Benchmarking circuits—and comparing mitigated results against ideal outcomes. Analysis of these benchmarking results produces perfor ...
+
+The benchmark circuits can be used using the following workflow.
```
A better introduction can be given to the snippet that follows: explain which benchmarking circuits you are generating, the fact that you are applying ZNE, that is run on a simulator, etc..

cosenal commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,194 @@
+
+## GHZ Circuits
+
+The {func}`.generate_ghz_circuit` create the GHZ states that are highly sensitive to noise. Thus, they make it easy to test error rates in entanglement creation and preservation, which is central for many quantum algorithms.
```
Could define what a GHZ state is, or add a reference. In general, don't assume much knowledge, many user come to Mitiq docs for learning :)

cosenal commented on docs/source/guide/benchmarks.md:
````diff
@@ -0,0 +1,194 @@
+print(circuit)
+```
+
+## Quantum Volume Circuits
````
I would sort the different kind of circuit in a more logical way, e.g., QV circuit together with Mirror-QV circuit, or GHZ circuits together with W circuits, etc..

Shivansh20128 replied:
Could you let me know the order in which you want them sorted? I don't have a deep level understanding about them so I am not sure what order they should be in...

cosenal replied:
It just felt a bit odd to see QV circuits apart from Mirror-QV circuits in the list. Maybe you can just put those two close to each other together for now. We can always reorganize the list in a future PR.

Shivansh20128 replied:
Done!

cosenal commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,194 @@
+
+## Mirror Quantum Volume Circuits
+
+The {func}`.generate_mirror_qv_circuit` is designed to test `quantum volume`, a metric combining circuit depth, number of qubits, and fidelity. These circuits check whether error mitigation techniques help achieve higher effective quantum volumes on noisy devices.
```
```suggestion
The {func}`.generate_mirror_qv_circuit` is designed to test [Quantum Volume](https://en.wikipedia.org/wiki/Quantum_volume), a metric combining circuit depth, number of qubits, and fidelity. These circuits check whether error mitigation techniques help achieve higher effective quantum volumes on noisy devices.
```

cosenal commented on docs/source/guide/benchmarks.md:
````diff
@@ -0,0 +1,194 @@
+```{code-cell} ipython3
+from mitiq.benchmarks import generate_mirror_qv_circuit
+
+circuit= generate_mirror_qv_circuit(num_qubits=7, depth=2)
````
nit: I don't think we strictly format tutorials/user guides with a linter, but it's still nice to have them formatted nicely.

```suggestion
circuit = generate_mirror_qv_circuit(num_qubits=7, depth=2)
```

cosenal commented on docs/source/guide/benchmarks.md:
````diff
@@ -0,0 +1,194 @@
+The {func}`.generate_qpe_circuit` is used to the measure eigenvalues of unitary operators. Since accurate phase estimation requires precise control over operations, these circuits test the mitigation techniques’ ability to handle small noise effects over multiple gate sequences.
+
+```{code-cell} ipython3
+from mitiq.benchmarks import generate_w_circuit
````
wrong function? 🤔

Shivansh20128 replied:
Yes. Sorry about this one. Don't know how this happened. Will correct this in the next commit.

cosenal reviewed (commented):
Overall this is what we had in mind. 
I agree with @purva-thakre that a snippet on how to use _each_ circuit type would make the user guide too heavy, so let's leave only the small demo at the top.
I left some comments throughout the PR, looking forward to the next iteration.

Shivansh20128 commented:
> Overall this is what we had in mind. I agree with @purva-thakre that a snippet on how to use _each_ circuit type would make the user guide too heavy, so let's leave only the small demo at the top.

@cosenal Just to make sure, you want to remove the all the code snippets from the circuits and leave just the one at the top? 
```{code-cell} ipython3
from mitiq.benchmarks import generate_ghz_circuit

circuit = generate_ghz_circuit(n_qubits=7)
```

So shall I remove all the blocks like these?

purva-thakre commented on docs/source/guide/benchmarks.md:
```diff
@@ -0,0 +1,133 @@
+  name: python3
+---
+
+# Benchmarks
```
```suggestion
# Benchmarking Circuits
```

Could you please relabel the file as `docs/source/guide/benchmarking_circuits.md`

cosenal commented:
> > Overall this is what we had in mind. I agree with @purva-thakre that a snippet on how to use _each_ circuit type would make the user guide too heavy, so let's leave only the small demo at the top.
> 
> @cosenal Just to make sure, you want to remove the all the code snippets from the circuits and leave just the one at the top?
> 
> ```
> from mitiq.benchmarks import generate_ghz_circuit
> 
> circuit = generate_ghz_circuit(n_qubits=7)
> ```
> 
> So shall I remove all the blocks like these?

No, those small snippets on how to generate the circuits are good. What I meant is that we shouldn't have snippets running error mitigation for each class of circuits.

purva-thakre commented:
What I had in mind for how each section describing a benchmarking circuit might look like:

![image](https://github.com/user-attachments/assets/340d48c8-9d7b-415f-adc2-aedfc6c9ca6e)

We can discuss this during the community call. 

[w_state_outline.txt](https://github.com/user-attachments/files/17873223/w_state_outline.txt)

purva-thakre commented on docs/source/guide/benchmarking-circuits.md:
````diff
@@ -0,0 +1,170 @@
+# Benchmarking Circuits
+
+Mitiq benchmarks error mitigation techniques by evaluating improvements in metrics such as state fidelity (the closeness of the mitigated quantum state to the ideal state), output probability distributions, and logical error rates. The benchmarking process involves running diverse circuit types—such as GHZ, Mirror, Quantum Volume, and Randomized Benchmarking circuits—and comparing mitigated resul ...
+
+The following workflow demonstrates how to use benchmark circuits in Mitiq. In this example, we generate a GHZ circuit using Mitiq’s benchmarking tools and apply Zero Noise Extrapolation (ZNE) to mitigate errors introduced by depolarizing noise. The workflow is run on a simulator, where we compare results from an ideal circuit, a noisy circuit, and a mitigated circuit to evaluate the impact of er ...
+
+```{code-cell} ipython3
+import cirq
+from mitiq import benchmarks, zne
+
+def execute(circuit, noise_level=0.005):
+    """Returns Tr[ρ |0⟩⟨0|] where ρ is the state prepared by the circuit
+    with depolarizing noise."""
+    noisy_circuit = circuit.with_noise(cirq.depolarize(p=noise_level))
+    return (
+        cirq.DensityMatrixSimulator()
+        .simulate(noisy_circuit)
+        .final_density_matrix[0, 0]
+        .real
+    )
+
+circuit = benchmarks.generate_ghz_circuit(n_qubits=7) # Call the required benchmark circuit function here
+print(circuit.final_state_vector()) # Shows the ideal circuit state
+print(circuit)
+
+true_value = execute(circuit, noise_level=0.0)      # Ideal quantum computer
+noisy_value = execute(circuit)                      # Noisy quantum computer
+zne_value = zne.execute_with_zne(circuit, execute)  # Noisy quantum computer + Mitiq
+
+print(f"Error w/o  Mitiq: {abs((true_value - noisy_value) / true_value):.3f}")
+print(f"Error w Mitiq:    {abs((true_value - zne_value) / true_value):.3f}")
+```
````
``` suggestion

```

I think it might be better to remove this block and the corresponding paragraph as the focus should be on benchmarking circuits. You can link other pages in the documentation that use the benchmarking circuits in QEM workflows.

purva-thakre commented on docs/source/guide/benchmarking-circuits.md:
```diff
@@ -0,0 +1,170 @@
+
+## Mirror Circuits
+
+The {func}`.generate_mirror_circuit` involves running a quantum circuit forward and then “mirroring” it (applying the reverse operations). Ideally, this results in returning the system to the initial state, so they’re great for testing if the noise mitigation is effective in preserving information through complex sequences.
```
It might be better to show an example code block.

purva-thakre replied:
Most of the benchmarking circuit docstrings have a link to some reference that was used to define the function. 

https://mitiq.readthedocs.io/en/stable/apidoc.html#module-mitiq.benchmarks.mirror_circuits

The W-state outline txt file contains an example of how to use {cite} for the references listed in `docs/source/refs.bib`

purva-thakre reviewed (changes requested):
I suggested some changes. @Shivansh20128 let me know if you have the time to work on these. 

If not, I can take over.

Shivansh20128 commented:
> I suggested some changes. @Shivansh20128 let me know if you have the time to work on these.

Okay. I will make the changes.

Shivansh20128 commented:
Hi @purva-thakre , I have added an example code block for the `generate_mirror_circuit`.
 
I have also added references for the benchmarking circuits wherever I could find them in the API-doc.

purva-thakre commented on docs/source/guide/benchmarking-circuits.md:
````diff
@@ -0,0 +1,149 @@
+```{code-cell} ipython3
+from mitiq.benchmarks import generate_ghz_circuit
+
+circuit = generate_ghz_circuit(n_qubits=7)
````
```suggestion
circuit = generate_ghz_circuit(n_qubits=7)

print(circuit)
```

Apply the same to other circuits on this page.

Shivansh20128 replied:
Done!

cosenal commented on docs/source/guide/benchmarking-circuits.md:
```diff
@@ -0,0 +1,167 @@
+
+## Mirror Quantum Volume Circuits
+
+The {func}`.generate_mirror_qv_circuit`, as defined in {cite}`Amico_2023_arxiv`, is designed to test [Quantum Volume](https://en.wikipedia.org/wiki/Quantum_volume), a metric combining circuit depth, number of qubits, and fidelity. These circuits check whether error mitigation techniques help achieve higher effective quantum volumes on noisy devices.
```
This description sounds the same as the QV one; instead, it should be noticed that here _mirror_ quantum volume circuits are generated with the same definition of _mirroring_ as above.

Shivansh20128 replied:
Done!

cosenal reviewed (approved):
Left a small comment. The rest looks good to me.
